txp/Pos/Aeson/Txp.hs

```diff
@@ -0,0 +1,99 @@
+
+module Pos.Aeson.Txp where
+
+import           Universum
+
+import           Data.Aeson           (FromJSON (..), FromJSONKey (..),
+                                       FromJSONKeyFunction (..), ToJSON (toJSON),
+                                       ToJSONKey (..), object, withObject, (.:), (.=))
+import           Data.Aeson.TH        (defaultOptions, deriveJSON)
+import           Data.Aeson.Types     (toJSONKeyText)
+import qualified Data.Text            as T
+import           Formatting           (build, int, sformat, (%))
+import qualified Serokell.Util.Base16 as B16
+import           Serokell.Util.Base64 (JsonByteString (..))
+
+import           Pos.Aeson.Core       ()
+import           Pos.Aeson.Crypto     ()
+import           Pos.Core             (coinToInteger, decodeTextAddress, integerToCoin)
+import           Pos.Crypto           (decodeAbstractHash, hashHexF)
+import           Pos.Txp.Core         (Tx, TxAux, TxIn (..), TxInWitness (..), TxOut (..),
+                                       TxOutAux, TxSigData)
+import           Pos.Util.Util        (eitherToFail)
+
+txInFromText :: Text -> Either Text TxIn
+txInFromText t = case T.splitOn "_" t of
+    ["TxInUtxo", h, idx]     -> TxInUtxo <$> decodeAbstractHash h <*> readEither idx
+    ["TxInUnknown", tag, bs] -> TxInUnknown <$> readEither tag <*> B16.decode bs
+    _                        -> fail $ toString $ "Invalid TxIn " <> t
+
+txInToText :: TxIn -> Text
+txInToText TxInUtxo {..} = sformat ("TxInUtxo_"%hashHexF%"_"%int) txInHash txInIndex
+txInToText (TxInUnknown tag bs) = sformat ("TxInUnknown_"%int%"_"%B16.base16F) tag bs
+
+instance FromJSON TxIn where
+    parseJSON v = eitherToFail =<< txInFromText <$> parseJSON v
+
+instance ToJSON TxIn where
+    toJSON = toJSON . txInToText
+
+instance FromJSONKey TxIn where
+    fromJSONKey = FromJSONKeyTextParser (eitherToFail . txInFromText)
+
+instance ToJSONKey TxIn where
+    toJSONKey = toJSONKeyText txInToText
+
+instance FromJSON TxOut where
+    parseJSON = withObject "TxOut" $ \o -> do
+        txOutValue   <- eitherToFail . integerToCoin =<< o .: "coin"
+        txOutAddress <- eitherToFail . decodeTextAddress =<< o .: "address"
+        return $ TxOut {..}
+
+instance ToJSON TxOut where
+    toJSON TxOut{..} = object [
+        "coin"    .= coinToInteger txOutValue,
+        "address" .= sformat build txOutAddress ]
+
+instance ToJSON TxInWitness where
+    toJSON = \case
+        PkWitness{..} -> object
+            [ "tag" .= ("PkWitness" :: Text)
+            , "twKey" .= twKey
+            , "twSig" .= twSig
+            ]
+        ScriptWitness{..} -> object
+            [ "tag" .= ("ScriptWitness" :: Text)
+            , "twValidator" .= twValidator
+            , "twRedeemer" .= twRedeemer
+            ]
+        RedeemWitness{..} -> object
+            [ "tag" .= ("RedeemWitness" :: Text)
+            , "twRedeemKey" .= twRedeemKey
+            , "twRedeemSig" .= twRedeemSig
```

Comment:
Tbh, I don't get why you decided to prepend `tw` prefixes, they look like unnecessary extra verbosity to me :/

Reply:
I concur!